text/0580-destroyables.md

```diff
@@ -0,0 +1,385 @@
+- Attempting to register the same destructor multiple times should throw an
+  error.
+
+#### `unregisterDestructor`
+
+Receives a destroyable and a destructor function, and de-associates the
+destructor from the destroyable.
+
+```js
+import { unregisterDestructor } from '@ember/destroyable';
+
+class Modal extends Component {
+  @service modals;
+
+  constructor() {
+    this.modals.add(this);
+
+    this.modalDestructor = registerDestructor(this, () => this.modals.remove(this));
+  }
+
+  @action pinModal() {
+    unregisterDestructor(this, this.modalDestructor);
+  }
+}
+```
+
+- Calling `unregisterDestructor` on a destroyed object should throw an error.
+- Calling `unregisterDestructor` with a destructor that is not associated with
+  the object should throw an error.
+
+#### `destroy`
+
+`destroy` initiates the destruction of a destroyable object. It runs all
+associated destructors, and then destroys all children recursively.
+
+```js
+let obj = {};
+
+registerDestructor(obj, () => console.log('destroyed!'));
+
+destroy(obj); // this will schedule the destructor to be called
+
+// ...some time later, during scheduled destruction
+
+// destroyed!
+```
+
+Destruction via `destroy()` follows these steps:
+
+1. Mark the destroyable such that `isDestroying(destroyable)` returns `true`
+2. Schedule calling the destroyable's destructors
+3. Call `destroy()` on each of the destroyable's associated children
+4. Schedule setting destroyable such that `isDestroyed(destroyable)` returns `true`
+
+This algorithm results in the entire tree of destroyables being first marked as
+destroying, then having all of their destructors called, and finally all being
+marked as `isDestroyed`. There won't be any in between states where some items
+are marked as `isDestroying` while destroying, while others are not.
+
+Calling `destroy` multiple times on the same destroyable is safe. It will not
+throw an error, and will not take any further action.
+
+Calling `destroy` with a destroyable that has no destructors or associated children
+will not throw an error, and will do nothing.
+
+#### `isDestroying`
+
+Receives a destroyable, and returns `true` if the destroyable has begun
+destroying. Otherwise returns false.
+
+```js
+let obj = {};
+
+isDestroying(obj); // false
+destroy(obj);
+isDestroying(obj); // true
+```
+
+#### `isDestroyed`
+
+Receives a destroyable, and returns `true` if the destroyable has finished
+destroying. Otherwise returns false.
+
+```js
+let obj = {};
+
+isDestroyed(obj); // false
+destroy(obj);
+
+// ...sometime later, after scheduled destruction
+
+isDestroyed(obj); // true
+```
+
+#### `assertDestroyablesDestroyed`
+
+This function asserts that all objects which have associated destructors or
+associated children have been destroyed at the time it is called. It is meant to
+be a low level hook that testing frameworks like `ember-qunit` and `ember-mocha`
+can use to hook into and validate that all destroyables have in fact been
+destroyed.
```

Comment:
It would be helpful to know which objects did not finish destroying in this case. The objets could be exposed as a property on the error.

Reply:
In my naive polyfill implementation, this function leaks state / failures between test runs, as there is no way to reset the hidden module state: https://github.com/buschtoens/ember-destroyable-polyfill/issues/2

```ts
test('should fail', function () {
  const obj = {};
  registerDestructor(obj, () => {});
  // run(() => destroy(obj));
  assertDestroyablesDestroyed();
});

test('should pass, but fails', function () {
  const obj = {};
  registerDestructor(obj, () => {});
  run(() => destroy(obj));
  assertDestroyablesDestroyed();
});
```

How are `ember-qunit` / `ember-mocha` supposed to reset the module state in `@ember/destroyable` after each test run? `require.unsee('@ember/destroyable')`? Or is this actually a design flaw?

Reply:
`assertDestroyablesDestroyed` could probably do the cleanup during test builds. We would have to track some extra state in DEBUG builds, use a normal Map instead of a WeakMap, but that's something we've done before.

Reply:
I've commented on the issue linked above (in the polyfill) to explain how this should work. I don't think this is a gap in the public APIs proposed (but the assertion would leverage internal private APIs).